### PR TITLE
Fixes #5325 - how to create help using PlatyPS

### LINKED ISSUE #5325: How to create help for Modules
# Documentation Issue
Current [article on Portable Modules](https://docs.microsoft.com/en-us/powershell/scripting/learn/writing-portable-modules?view=powershell-7) does not mention how to provide the help section for the custom portable modules.

## Context of the issue

- URL to the article (Docs or GitHub URL): https://docs.microsoft.com/en-us/powershell/scripting/learn/writing-portable-modules?view=powershell-7
- Issue affects the following content (check all that apply):

  **Cmdlet reference & about_ topics**
  - [X] Version 7.x preview content
  - [X] Version 7.0 content
  - [X] Version 6 content
  - [X] Version 5.1 content

## Detailed description of the issue
In the old days we were encouraged to document scripts/modules [using comments based help](https://docs.microsoft.com/en-us/powershell/module/microsoft.powershell.core/about/about_comment_based_help?view=powershell-7). In the new way f binary modules we can't use same method since C# comments use different structure. Can we please document how to provide documentations for binary modules?

### PR DESCRIPTION
**Pull request by sdwheeler**
# PR Summary

Fixes [AB#1751480](https://dev.azure.com/mseng/677da0fb-b067-4f77-b89b-f32c12bb8617/_workitems/edit/1751480) - Fixes #5325 - how to create help using PlatyPS

## PR Context

Select the area of the Table of Contents containing the documents being changed.

**Conceptual content**
- [ ] Overview and Install
- [ ] Learning PowerShell
  - [ ] PowerShell 101
  - [ ] Deep dives
  - [ ] Remoting
- [ ] Release notes (What's New)
- [ ] Windows PowerShell
  - WMF, ISE, release notes, etc.
- [ ] DSC articles
- [ ] Community resources
- [ ] Sample scripts
- [ ] Gallery articles
- [x] Scripting and development
  - [ ] Legacy SDK

**Cmdlet reference & about_ topics**
- [ ] Version 7.1 preview content
- [ ] Version 7.0 content
- [ ] Version 6 content
- [ ] Version 5.1 content

## PR Checklist

- [x] I have read the [contributors guide][contrib] and followed the style and process guidelines
- [x] PR has a meaningful title
- [x] PR is targeted at the _staging_ branch
- [x] All relevant versions updated
- [x] Includes content related to issues and PRs - see [Closing issues using keywords][key].
- [x] This PR is ready to merge and is not **Work in Progress**
  - If the PR is work in progress, please add the prefix `WIP:` or `[WIP]` to the beginning of the
    title and remove the prefix when the PR is ready.

[contrib]: https://docs.microsoft.com/powershell/scripting/community/contributing/overview
[key]: https://help.github.com/en/articles/closing-issues-using-keywords

### CODE REVIEW

File: reference/docs-conceptual/dev-cross-plat/create-help-using-platyps.md [hunk 1]
```diff
@@ -0,0 +1,294 @@
...
+1. Add a **HelpInfoURI** key to your module manifest
+1. Create a HelpInfo XML file
+1. Create CAB files
+1. Upload your files
+
+For more information, see [Supporting Updateable Help: Step-by-step][step-by-step].
+
+PlatyPS makes some of these steps easier.
+
+The **HelpInfoURI** is a URL that points to location where your help are hosted on the internet.
```
Comment (chasewilson) [inline reference/docs-conceptual/dev-cross-plat/create-help-using-platyps.md:229]:
```suggestion
The **HelpInfoURI** is a URL that points to location where your help files are hosted on the internet.
```

File: reference/docs-conceptual/dev-cross-plat/create-help-using-platyps.md [hunk 2]
```diff
@@ -0,0 +1,294 @@
...
+This value is configured in the module manifest. `Update-Help` reads the module manifest to get this
+URL and download the updateable help content. This URL should only point to the folder location and
```
Comment (chasewilson) [inline reference/docs-conceptual/dev-cross-plat/create-help-using-platyps.md:231]:
```suggestion
URL and downloads the updateable help content. This URL should only point to the folder location and
```

File: reference/docs-conceptual/dev-cross-plat/create-help-using-platyps.md [hunk 3]
```diff
@@ -0,0 +1,294 @@
...
+> `Update-Help` cannot construct the correct file paths to download the content. Also, most
+> HTTP-based files services are case-sensitive. It is important that the module metadata in the
+> HelpInfo XML file contains the proper letter case.
+
+The `New-ExternalHelp` cmdlet creates the HelpInfo XML file in the output folder. The HelpInfo XML
+file is built using YAML metadata contained in the module Markdown files (`ModuleName.md`).
+
+The `New-ExternalHelpCab` cmdlet creates ZIP and CAB files containing the MAML and
+`about_*.help.txt` files you compiled. CAB files are compatible with PowerShell 5.1 and older. Newer
+versions of PowerShell can use CAB or ZIP files.
```
Comment (DCtheGeek) [inline reference/docs-conceptual/dev-cross-plat/create-help-using-platyps.md:246]: These last sentences read a bit funny. Can we just say CAB files are compatible with all versions of PowerShell and 6.0+ can use ZIP files or something along those lines?

File: reference/docs-conceptual/dev-cross-plat/create-help-using-platyps.md [hunk 4]
```diff
@@ -0,0 +1,294 @@
...
+The **HelpInfoURI** is a URL that points to location where your help are hosted on the internet.
+This value is configured in the module manifest. `Update-Help` reads the module manifest to get this
+URL and download the updateable help content. This URL should only point to the folder location and
+not to individual files. `Update-Help` constructs the filenames to download based on other
+information from the module manifest and the HelpInfo XML file.
+
+> [!IMPORTANT]
+> The **HelpInfoURI** must end with a forward-slash character (`/`). Without that character,
+> `Update-Help` cannot construct the correct file paths to download the content. Also, most
+> HTTP-based files services are case-sensitive. It is important that the module metadata in the
```
Comment (chasewilson) [inline reference/docs-conceptual/dev-cross-plat/create-help-using-platyps.md:238]:
```suggestion
> HTTP-based file services are case-sensitive. It is important that the module metadata in the
```

File: reference/docs-conceptual/dev-cross-plat/create-help-using-platyps.md [hunk 5]
```diff
@@ -0,0 +1,294 @@
+---
+title: Create XML-based help using PlatyPS
+description: Create XML-based help the easy way using PlatyPS.
```
[the remark this answers is not in the document]
Comment (sdwheeler) [reply reference/docs-conceptual/dev-cross-plat/create-help-using-platyps.md:3]: meh

File: reference/docs-conceptual/dev-cross-plat/create-help-using-platyps.md [hunk 6]
```diff
@@ -0,0 +1,294 @@
...
+ms.date: 06/25/2020
```
[the remark this answers is not in the document]
Comment (sdwheeler) [reply reference/docs-conceptual/dev-cross-plat/create-help-using-platyps.md:4]: fixed

File: reference/docs-conceptual/dev-cross-plat/create-help-using-platyps.md [hunk 7]
```diff
@@ -0,0 +1,294 @@
...
+The legacy PowerShell SDK documentation covers the details of creating help for PowerShell
+cmdlets packaged into modules. However, PowerShell does not provide any tools for creating the
+XML-based help. The SDK documentation explains the structure of MAML help. but leaves you the
+task of creating the complex, and deeply nested, MAML content by hand.
+
+This is where the [PlatyPS][] module can help.
+
+## What is PlatyPS?
+
+PlatyPS is an [open-source][platyps-repo] tool that was created a _hackathon_ project to make the
```
[the remark this answers is not in the document]
Comment (sdwheeler) [reply reference/docs-conceptual/dev-cross-plat/create-help-using-platyps.md]: your comment is outdated

File: reference/docs-conceptual/dev-cross-plat/create-help-using-platyps.md [hunk 8]
```diff
@@ -0,0 +1,294 @@
...
+individual parameters for each cmdlet in your module. The PlatyPS creates structured [Markdown][]
+files that contain the syntax information. It can't create descriptions or provide examples.
+
+PlatyPS creates placeholders for you to fill in descriptions and examples. After adding the required
+information, use PlatyPS compiles the Markdown files into MAML files. PowerShell's help system also
+allows for plain-text conceptual help files (about topics). PlatyPS has a cmdlet to create a
+structured Markdown template for a new _about_ file, but these `about_*.help.txt` files must be
+maintained manually.
+
+You can include the MAML and TXT help files with your module. Or you can use PlatyPS to compile the
```
[the remark this answers is not in the document]
Comment (sdwheeler) [reply reference/docs-conceptual/dev-cross-plat/create-help-using-platyps.md:32]: Fixed

File: reference/docs-conceptual/dev-cross-plat/create-help-using-platyps.md [hunk 9]
```diff
@@ -0,0 +1,294 @@
...
+help files into a CAB package that can be hosted for updateable help.
```
[the remark this answers is not in the document]
Comment (sdwheeler) [reply reference/docs-conceptual/dev-cross-plat/create-help-using-platyps.md:33]: Well understood.

File: reference/docs-conceptual/dev-cross-plat/create-help-using-platyps.md [hunk 10]
```diff
@@ -0,0 +1,294 @@
...
+First you must install PlatyPS from the PowerShell Gallery.
+
+```powershell
+Install-Module platyps -Force
+Import-Module platyps
+```
+
+The following flowchart outlines the process for creating or updating PowerShell reference content.
+
+:::image type="content" source="./media/create-help-using-platyps/cmdlet-ref-flow-v2.png" alt-text="Cmdlet reference workflow":::
```
[the remark this answers is not in the document]
Comment (sdwheeler) [reply reference/docs-conceptual/dev-cross-plat/create-help-using-platyps.md:46]: fixed

File: reference/docs-conceptual/dev-cross-plat/create-help-using-platyps.md [hunk 11]
```diff
@@ -0,0 +1,294 @@
...
+       UpdateInputOutput = $true
+       ExcludeDontShow = $true
+       LogPath = <path to store log file>
+       Encoding = 'UTF8BOM'
+   }
+   Update-MarkdownHelpModule @parameters
+   ```
+
+   `Update-MarkdownHelpModule` updates any cmdlet and module Markdown files in the specified folder.
+   It does not update any `about_*.md` files. The module file (`ModuleName.md`) will receive any new
```
[the remark this answers is not in the document]
Comment (sdwheeler) [reply reference/docs-conceptual/dev-cross-plat/create-help-using-platyps.md]: already fixed

File: reference/docs-conceptual/dev-cross-plat/create-help-using-platyps.md [hunk 12]
```diff
@@ -0,0 +1,294 @@
...
+- [Editing reference articles](/powershell/scripting/community/contributing/editing-cmdlet-ref)
+
+PlatyPS has a specific schema that is uses for cmdlet reference. That schema only allows certain
+Markdown blocks in specific sections of the document. If you put the unsupported content in wrong
+location, the PlatyPS build step fails. See the [schema][] documentation in the PlatyPS repository.
+
+For a complete example of well-formed cmdlet reference, see [Get-Item][].
+
+After providing the required content for each of your cmdlets, you need to make sure that you update
+the module landing page. Verify your module has the correct GUID and Help Link in the YAML
```
[the remark this answers is not in the document]
Comment (sdwheeler) [reply reference/docs-conceptual/dev-cross-plat/create-help-using-platyps.md:146]: fixed

File: reference/docs-conceptual/dev-cross-plat/create-help-using-platyps.md [hunk 13]
```diff
@@ -0,0 +1,294 @@
...
+front-matter of the `<module-name>.md` file. Add any missing metadata.
```
[the remark this answers is not in the document]
Comment (sdwheeler) [reply reference/docs-conceptual/dev-cross-plat/create-help-using-platyps.md:147]: fixed

File: reference/docs-conceptual/dev-cross-plat/create-help-using-platyps.md [hunk 14]
```diff
@@ -0,0 +1,294 @@
...
+
+1. Find an internet site to host your help files
+1. Add a **HelpInfoURI** key to your module manifest
+1. Create a HelpInfo XML file
+1. Create CAB files
+1. Upload your files
+
+For more information, see [Supporting Updateable Help: Step-by-step][step-by-step].
+
+PlatyPS makes some of these steps easier.
```
[the remark this answers is not in the document]
Comment (sdwheeler) [reply reference/docs-conceptual/dev-cross-plat/create-help-using-platyps.md:227]: fixed

File: reference/docs-conceptual/dev-cross-plat/create-help-using-platyps.md (hunk 3 above)
Comment (sdwheeler) [reply reference/docs-conceptual/dev-cross-plat/create-help-using-platyps.md:246]: Fixed

File: reference/docs-conceptual/dev-cross-plat/create-help-using-platyps.md [hunk 15]
```diff
@@ -0,0 +1,294 @@
+---
+title: Create XML-based help using PlatyPS
+description: Create XML-based help the easy way using PlatyPS.
+ms.date: 06/25/2020
+---
+# Create XML-based help using PlatyPS
+
+When creating a PowerShell module, it is always recommended that you include detailed help for the
+cmdlets you created. If your cmdlets are implemented in compiled code, you must use the XML-based
```
[the remark this answers is not in the document]
Comment (sdwheeler) [reply reference/docs-conceptual/dev-cross-plat/create-help-using-platyps.md:9]: fixed

File: reference/docs-conceptual/dev-cross-plat/create-help-using-platyps.md [hunk 16]
```diff
@@ -0,0 +1,294 @@
...
+help. This XML format is known as the Microsoft Assistance Markup Language (MAML).
```
[the remark this answers is not in the document]
Comment (sdwheeler) [reply reference/docs-conceptual/dev-cross-plat/create-help-using-platyps.md:10]: going to leave this as is

File: reference/docs-conceptual/dev-cross-plat/create-help-using-platyps.md [hunk 17]
```diff
@@ -0,0 +1,294 @@
...
+
+The legacy PowerShell SDK documentation covers the details of creating help for PowerShell
+cmdlets packaged into modules. However, PowerShell does not provide any tools for creating the
+XML-based help. The SDK documentation explains the structure of MAML help. but leaves you the
```
[the remark this answers is not in the document]
Comment (sdwheeler) [reply reference/docs-conceptual/dev-cross-plat/create-help-using-platyps.md]: fixed

File: reference/docs-conceptual/dev-cross-plat/create-help-using-platyps.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (sdwheeler) [reply reference/docs-conceptual/dev-cross-plat/create-help-using-platyps.md]: fixed

File: reference/docs-conceptual/dev-cross-plat/create-help-using-platyps.md [hunk 18]
```diff
@@ -0,0 +1,294 @@
...
+task of creating the complex, and deeply nested, MAML content by hand.
+
+This is where the [PlatyPS][] module can help.
+
+## What is PlatyPS?
+
+PlatyPS is an [open-source][platyps-repo] tool that was created a _hackathon_ project to make the
+creation and maintenance of MAML easier. PlatyPS documents the syntax of the parameter sets and the
```
[the remark this answers is not in the document]
Comment (sdwheeler) [reply reference/docs-conceptual/dev-cross-plat/create-help-using-platyps.md:22]: Fixed

File: reference/docs-conceptual/dev-cross-plat/create-help-using-platyps.md [hunk 19]
```diff
@@ -0,0 +1,294 @@
...
+individual parameters for each cmdlet in your module. The PlatyPS creates structured [Markdown][]
+files that contain the syntax information. It can't create descriptions or provide examples.
+
+PlatyPS creates placeholders for you to fill in descriptions and examples. After adding the required
+information, use PlatyPS compiles the Markdown files into MAML files. PowerShell's help system also
```
[the remark this answers is not in the document]
Comment (sdwheeler) [reply reference/docs-conceptual/dev-cross-plat/create-help-using-platyps.md:27]: Fixed

File: reference/docs-conceptual/dev-cross-plat/create-help-using-platyps.md (hunk 8 above)
[the remark this answers is not in the document]
Comment (sdwheeler) [reply reference/docs-conceptual/dev-cross-plat/create-help-using-platyps.md:32]: Fixed

File: reference/docs-conceptual/dev-cross-plat/create-help-using-platyps.md (hunk 9 above)
[the remark this answers is not in the document]
Comment (sdwheeler) [reply reference/docs-conceptual/dev-cross-plat/create-help-using-platyps.md:33]: stet

File: reference/docs-conceptual/dev-cross-plat/create-help-using-platyps.md [hunk 20]
```diff
@@ -0,0 +1,294 @@
...
+       Encoding = 'UTF8BOM'
+   }
+   New-MarkdownHelp @parameters
+
+   New-MarkdownAboutHelp -OutputFolder $OutputFolder -AboutName "topic_name"
+   ```
+
+   If the output folder does not exist, `New-MarkdownHelp` creates it. In this example,
+   `New-MarkdownHelp` creates a Markdown file for each cmdlet in the module. It also creates a file
+   named `ModuleName.md`. This file contains a list of the cmdlets contained in the module and
```
[the remark this answers is not in the document]
Comment (sdwheeler) [reply reference/docs-conceptual/dev-cross-plat/create-help-using-platyps.md:78]: Fixed

File: reference/docs-conceptual/dev-cross-plat/create-help-using-platyps.md [hunk 21]
```diff
@@ -0,0 +1,294 @@
...
+You need to add a synopsis, a description of the cmdlet, descriptions for each parameter, and
+at least one example.
+
+For detailed information about writing PowerShell content, see the following articles:
+
+- [PowerShell style guide](/powershell/scripting/community/contributing/powershell-style-guide)
+- [Editing reference articles](/powershell/scripting/community/contributing/editing-cmdlet-ref)
+
+PlatyPS has a specific schema that is uses for cmdlet reference. That schema only allows certain
+Markdown blocks in specific sections of the document. If you put the unsupported content in wrong
```
[the remark this answers is not in the document]
Comment (sdwheeler) [reply reference/docs-conceptual/dev-cross-plat/create-help-using-platyps.md:140]: Fixed

File: reference/docs-conceptual/dev-cross-plat/create-help-using-platyps.md (hunk 1 above)
Comment (sdwheeler) [reply reference/docs-conceptual/dev-cross-plat/create-help-using-platyps.md:229]: Fixed

File: reference/docs-conceptual/dev-cross-plat/create-help-using-platyps.md [hunk 22]
```diff
@@ -0,0 +1,294 @@
...
+
+You can verify the content with the [Get-HelpPreview][] cmdlet:
+
+```powershell
+Get-HelpPreview -Path "<ModuleName>-Help.xml"
+```
+
+The cmdlet reads the compiled MAML file and outputs the content in the same format you would receive
+from `Get-Help`. This allows you to inspect the results to verify that the Markdown files compiled
+correctly and produce the desired results. If you find error, re-edit the Markdown files and
```
[the remark this answers is not in the document]
Comment (sdwheeler) [reply reference/docs-conceptual/dev-cross-plat/create-help-using-platyps.md:196]: Fixed

File: reference/docs-conceptual/dev-cross-plat/create-help-using-platyps.md (hunk 4 above)
Comment (sdwheeler) [reply reference/docs-conceptual/dev-cross-plat/create-help-using-platyps.md:238]: Fixed

File: reference/docs-conceptual/dev-cross-plat/create-help-using-platyps.md [hunk 23]
```diff
@@ -0,0 +1,297 @@
...
+
+1. Import your new module into the session. Repeat this step for each module you need to document.
+
+   Run the following command to import your modules:
+
+   ```powershell
+   Import-Module <your module name>
+   ```
+
+1. Use PlatyPS to generate Markdown files for your module landing page and all associated cmdlets
```
Comment (theJasonHelmick) [inline reference/docs-conceptual/dev-cross-plat/create-help-using-platyps.md:58]: Do you think we should define what a Module Landing page is?

File: reference/docs-conceptual/dev-cross-plat/create-help-using-platyps.md [hunk 24]
```diff
@@ -0,0 +1,297 @@
...
+   - Updated parameter metadata
+   - Updated input and output type information
+
+   For more information, see [Update-MarkdownHelpModule][].
+
+## Edit the new or updated Markdown files
+
+PlatyPS documents the syntax of the parameter sets and the individual parameters. It can't create
+descriptions or provide examples. The specific areas where content is needed are contained in curly
+braces. For example: `{{*Fill in the Synopsis*}}`
```
Comment (theJasonHelmick) [inline reference/docs-conceptual/dev-cross-plat/create-help-using-platyps.md:129]: I wonder if an image of the file with highlights on {{*Fill in the ...}} would help visualize the edits needed.

Comment (sdwheeler) [reply reference/docs-conceptual/dev-cross-plat/create-help-using-platyps.md:129]: I will look into this.

File: reference/docs-conceptual/dev-cross-plat/create-help-using-platyps.md (hunk 23 above)
Comment (sdwheeler) [reply reference/docs-conceptual/dev-cross-plat/create-help-using-platyps.md:58]: I can expand on this.

**Pull request merged**